controller/composite/metacontroller.go

```diff
@@ -17,153 +17,81 @@ limitations under the License.
+func NewMetacontroller(k8s client.Client, resources *dynamicdiscovery.ResourceMap, dynClient *dynamicclientset.Clientset, dynInformers *dynamicinformer.SharedInformerFactory, mcInformerFactory mcinformers.SharedInformerFactory, mcClient mcclientset.Interface, numWorkers int) *Metacontroller {
 	mc := &Metacontroller{
+		k8s:          k8s,
 		resources:    resources,
 		mcClient:     mcClient,
 		dynClient:    dynClient,
 		dynInformers: dynInformers,
 
-		ccLister:         mcInformerFactory.Metacontroller().V1alpha1().CompositeControllers().Lister(),
-		ccInformer:       mcInformerFactory.Metacontroller().V1alpha1().CompositeControllers().Informer(),
 		revisionLister:   mcInformerFactory.Metacontroller().V1alpha1().ControllerRevisions().Lister(),
 		revisionInformer: mcInformerFactory.Metacontroller().V1alpha1().ControllerRevisions().Informer(),
 
-		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CompositeController"),
 		parentControllers: make(map[string]*parentController),
 
 		numWorkers: numWorkers,
 	}
 
-	mc.ccInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    mc.enqueueCompositeController,
-		UpdateFunc: mc.updateCompositeController,
-		DeleteFunc: mc.enqueueCompositeController,
-	})
-
 	return mc
 }
 
-func (mc *Metacontroller) Start() {
-	mc.stopCh = make(chan struct{})
-	mc.doneCh = make(chan struct{})
-
-	go func() {
-		defer close(mc.doneCh)
-		defer utilruntime.HandleCrash()
-
-		klog.InfoS("Starting CompositeController metacontroller")
-		defer klog.InfoS("Shutting down CompositeController metacontroller")
-
-		if !cache.WaitForNamedCacheSync("CompositeController", mc.stopCh, mc.ccInformer.HasSynced) {
-			return
-		}
-
-		// In the metacontroller, we are only responsible for starting/stopping
-		// the actual controllers, so a single worker should be enough.
-		for mc.processNextWorkItem() {
-		}
-	}()
-}
-
-func (mc *Metacontroller) Stop() {
-	// Stop metacontroller first so there's no more changes to controllers.
-	close(mc.stopCh)
-	mc.queue.ShutDown()
-	<-mc.doneCh
-
-	// Stop all controllers.
-	var wg sync.WaitGroup
-	for _, pc := range mc.parentControllers {
-		wg.Add(1)
-		go func(pc *parentController) {
-			defer wg.Done()
-			pc.Stop()
-		}(pc)
-	}
-	wg.Wait()
-}
-
-func (mc *Metacontroller) processNextWorkItem() bool {
-	key, quit := mc.queue.Get()
-	if quit {
-		return false
-	}
-	defer mc.queue.Done(key)
-
-	err := mc.sync(key.(string))
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("failed to sync CompositeController %q: %v", key, err))
-		mc.queue.AddRateLimited(key)
-		return true
-	}
-
-	mc.queue.Forget(key)
-	return true
-}
-
-func (mc *Metacontroller) sync(key string) error {
-	_, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		return err
-	}
-
+func (mc *Metacontroller) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
Does this need to be named as `Reconcile`?
Can we have this as `Sync` & rename `syncCompositeController` to `sync`
Or we continue having this as `Reconcile` & rename `syncCompositeController` to `reconcile`
It may be more appropriate to think these as separate PRs as well.
I was thinking more in terms of readability, idiomatic conventions wherever possible.

Reply:
I am voting to renaming `sync ` to `reconcileCompositeController` as it seems it is the term used among controllers

Reply:
The name needs to be `Reconcile` since it is the one defined by the interface. However I renamed the non-exported `sync` methods to `reconcileCompositeController` and `reconcileDecoratorController` respectively.

Reply:
Actually I think changing the name can be good, as `reconcile` is widely used term of syncing resources in operator world